lib/types.nix

```diff
@@ -594,6 +594,11 @@ rec {
               else throw "A submoduleWith option is declared multiple times with conflicting shorthandOnlyDefinesConfig values";
           };
         };
+      } // {
+        # Submodule-typed options get special treatment in order to facilitate
+        # certain migrations, such as the addition of freeformTypes onto
+        # existing option trees.
+        isSubmodule = true;
```

Comment:
I think for further extensibility it would make sense to abstract this into a function, so that e.g. `listOf` can define its own logic for things like `foo."*".bar`. This would replace `optionTreeToOption` (or perhaps `map optionTreeToOption`) in `lib/modules`.

Reply:
I agree, but I can't decide this without also deciding what the semantics of, say, `attrsOf` should be when it comes to option merging. It's ok to take small steps. I've added an explanation as an inline comment.

I've removed `isSubmodule` is it wasn't strictly necessary. This way we won't have to remove it when we do generalize this feature.